src/infra/core/src/remote_alias_resolver_impl.rs

```diff
@@ -0,0 +1,296 @@
+    ) -> Result<odf::DatasetName, ResolveAliasError> {
+        let result = if let Some(remote_dataset_name) =
+            RemoteAliasResolverApiHelper::fetch_remote_dataset_name(
+                remote_repo_url,
+                &dataset_handle.id,
+                access_token_maybe,
+            )
+            .await?
+        {
+            remote_dataset_name
+        } else {
+            dataset_handle.alias.dataset_name.clone()
+        };
+        Ok(result)
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[async_trait::async_trait]
+impl RemoteAliasResolver for RemoteAliasResolverImpl {
+    async fn resolve_remote_alias(
+        &self,
+        local_dataset_handle: &odf::DatasetHandle,
+        dataset_push_target_maybe: Option<odf::DatasetPushTarget>,
+        remote_alias_kind: RemoteAliasKind,
+    ) -> Result<RemoteAliasRef, ResolveAliasError> {
+        let repo_name: odf::RepoName;
+        let mut account_name = None;
+        let mut dataset_name = None;
+
+        if let Some(dataset_push_target) = &dataset_push_target_maybe {
+            match dataset_push_target {
+                odf::DatasetPushTarget::Alias(dataset_alias_remote) => {
+                    repo_name = dataset_alias_remote.repo_name.clone();
+                    account_name.clone_from(&dataset_alias_remote.account_name);
+                    dataset_name = Some(dataset_alias_remote.dataset_name.clone());
+                }
+                odf::DatasetPushTarget::Url(url_ref) => {
+                    return Ok(RemoteAliasRef::new(
+                        url_ref.clone(),
+                        None,
+                        dataset_name,
+                        account_name,
+                    ));
+                }
+                odf::DatasetPushTarget::Repository(repository_name) => {
+                    repo_name = repository_name.clone();
+                }
+            }
+        } else {
+            if let Some(remote_url) = self
+                .fetch_remote_url(local_dataset_handle, remote_alias_kind)
+                .await?
+            {
+                return Ok(RemoteAliasRef::new(
+                    remote_url,
+                    None,
+                    dataset_name,
+                    account_name,
+                ));
+            }
+            let remote_repo_names: Vec<_> = self.remote_repo_reg.get_all_repositories().collect();
+            if remote_repo_names.len() > 1 {
+                return Err(ResolveAliasError::AmbiguousRepository);
+            }
+            if let Some(repository_name) = remote_repo_names.first() {
+                repo_name = repository_name.clone();
+            } else {
+                return Err(ResolveAliasError::EmptyRepositoryList);
+            }
+        }
+        let remote_repo = self.remote_repo_reg.get_repository(&repo_name).int_err()?;
```

Comment:
If `dataset_push_target_maybe == Some(Repository("blah"))` and "blah" repository does not exist will we get an internal error here instead of `RepositoryNotFound`?

Reply:
Good catch, fixed